### PR TITLE
Fix: Order API Permissions

### LINKED ISSUE #4358: Ensure permissions in the orders API comply with the handbook
**I'm submitting a ...**  (check one with "x")
- [x] bug report
- [ ] feature request
- [ ] support request => Please do not submit support requests here, instead ask your query in out Gitter channel at https://gitter.im/fossasia/open-event-orga-server

**Current behavior:**

**Expected behavior:**

**Steps to reproduce:**

**Related code:**

```
insert any relevant code here else remove this section
```

**Other information:**

**System information:** 

```
Your operating system
```

```
output of `python --version`
```

### PR DESCRIPTION
**Pull request by poush**
Fixes #4358 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:

Update permissions according to dev handbook

### CODE REVIEW

File: app/api/orders.py [hunk 1]
```diff
@@ -116,24 +116,41 @@ def before_get_object(self, view_kwargs):
...
+        if not has_access('is_coorganizer_or_user_itself', event_id=order.event_id, user_id=order.user_id):
+            return ForbiddenException({'source': ''}, 'Access Forbidden')
+
+
     def before_update_object(self, order, data, view_kwargs):
+
+        if not has_access('is_admin'):
+            for element in data:
+                if element != 'status':
+                    del data[element]
```
Comment (shubham-padia) [inline app/api/orders.py]: Don't delete these, that makes them null. Just assign them their previous values again.

File: app/api/orders.py [hunk 2]
```diff
@@ -116,24 +116,41 @@ def before_get_object(self, view_kwargs):
...
+
         if data.get('status'):
             if has_access('is_coorganizer', event_id=order.event.id):
                 pass
             else:
                 raise ForbiddenException({'pointer': 'data/status'},
```
Comment (shubham-padia) [inline app/api/orders.py]: replace with,
```
if not has_acces..
raise ....
```

File: app/api/orders.py [hunk 3]
```diff
@@ -116,24 +116,41 @@ def before_get_object(self, view_kwargs):
...
                                          "To update status minimum Co-organizer access required")
 
-    decorators = (api.has_permission('is_coorganizer', fetch="event_id", fetch_as="event_id",
-                                     fetch_key_model="identifier", fetch_key_url="order_identifier", model=Order),)
+    def before_delete_object(self, order, view_kwargs):
+        if not has_access('is_coorganizer', event_id=order.event.id):
+            return ForbiddenException({'source': ''}, 'Access Forbidden')
```
Comment (shubham-padia) [inline app/api/orders.py]: `return` -> `raise`

**Comment by SaptakS** [comment]
@poush please fix conflicts as well. Thanks.

File: app/api/orders.py [hunk 4]
```diff
@@ -125,32 +125,47 @@ def before_get_object(self, view_kwargs):
...
 
+        order = safe_query(self, Order, 'identifier', view_kwargs['order_identifier'], 'order_identifier')
+
+        if not has_access('is_coorganizer_or_user_itself', event_id=order.event_id, user_id=order.user_id):
+            return ForbiddenException({'source': ''}, 'Access Forbidden')
+
+
     def before_update_object(self, order, data, view_kwargs):
         if data.get('cancel_note'):
             del data['cancel_note']
```
Comment (shubham-padia) [inline app/api/orders.py]: Why delete `cancel_note` :sweat_smile: ?

File: app/api/orders.py [hunk 5]
```diff
@@ -125,32 +125,47 @@ def before_get_object(self, view_kwargs):
             attendee = safe_query(self, TicketHolder, 'id', view_kwargs['attendee_id'], 'attendee_id')
             view_kwargs['order_identifier'] = attendee.order.identifier
 
+        order = safe_query(self, Order, 'identifier', view_kwargs['order_identifier'], 'order_identifier')
+
+        if not has_access('is_coorganizer_or_user_itself', event_id=order.event_id, user_id=order.user_id):
+            return ForbiddenException({'source': ''}, 'Access Forbidden')
+
+
     def before_update_object(self, order, data, view_kwargs):
```
Comment (shubham-padia) [inline app/api/orders.py]: According to the handbook, only status should be allowed to change by the organizer, once the order has been created, please implement that

File: app/api/orders.py (hunk 4 above)
Comment (poush) [inline app/api/orders.py]: must have missed it in conflict resolving. Basically was trying to delete cancel_note if the status is not set to cancelled.

Comment (shubham-padia) [inline app/api/orders.py]: Even if the status is not cancelled, don't delete cancel_note, let it be there

Comment (poush) [inline app/api/orders.py]: Okay, Done

**Pull request merged**